### PR TITLE
fix edge case for color NCS 0580-G76Y (#5)

### PR DESCRIPTION
**Pull request by milahu**

### CODE REVIEW

File: ncscolor.js [hunk 1]
```diff
@@ -116,7 +116,9 @@
         top = Gc
       } else if (Bc > Rc && Bc > Gc) {
         top = Bc
-      } else {
+      } else if (Rc == Gc) {
```
Comment (m90) [inline ncscolor.js]: So this library has a pretty twisted history: some time around 2008 I needed this functionality in a Flash project and ported some random Java Code I found on a forum. At some later point I ported the Flash project to JavaScript and this repo got created. This means the code is as twisted as it is right now.

Looking at this with a little more distance I wonder if this pretty complicated procedure couldn't simply be:

```
top = Math.max(Rc, Bc, Gc)
```

because we basically just want the biggest of these values? Does this make sense to you as well?

**Review by m90 — COMMENTED** [review]
Thanks for following up on this. Do you think you could add a test for `NCS 0580-G76Y` so this does not regress?

**Comment by milahu** [comment]
now there is a conflict with [w3color.js](https://www.w3schools.com/colors/colors_converter.asp?color=ncs(0580-G76Y))

```
$ node compare-ncscolor-vs-w3color.js
mismatch for NCS 0580-G76Y: w3color rgb(255,255,51) vs rgb(255,255,36) ncsColor
```

w3color neighbors are

```
ncs(0580-G74Y) rgb(251, 255, 36)
ncs(0580-G75Y) rgb(253, 255, 36)
ncs(0580-G76Y) rgb(255, 255, 51)
ncs(0580-G77Y) rgb(254, 253, 36)
ncs(0580-G78Y) rgb(255, 251, 36)
```

looks like w3color is wrong : D

https://github.com/milahu/compare-ncscolor-vs-w3color

**Comment by m90** [comment]
> now there is a conflict with w3color.js

What is "now" exactly?

**Comment by milahu** [comment]
> What is "now" exactly?

with my patch

**Comment by m90** [comment]
As converting NCS to RGB will always be an approximation of sorts I would expect this to happen on a lot of values. Interesting though that w3color (which I have never heard of until today) seems to have problems on the same color as well.

I would assume both libs are fine as long as their output is consistent when compared to itself.

**Comment by milahu** [comment]
> I would assume both libs are fine as long as their output is consistent when compared to itself.

look at the blue-value around 0580-G76Y:

```
               w3color           ncs-color
ncs(0580-G74Y) rgb(251, 255, 36) rgb(251, 255, 36)
ncs(0580-G75Y) rgb(253, 255, 36) rgb(253, 255, 36)
ncs(0580-G76Y) rgb(255, 255, 51) rgb(255, 255, 36)
ncs(0580-G77Y) rgb(254, 253, 36) rgb(254, 253, 36)
ncs(0580-G78Y) rgb(255, 251, 36) rgb(255, 251, 36)
```

**Comment by m90** [comment]
Yeah it seems to fall for some similar problem on the same color. Probably some rare NCS Edge Case.

Do you think you could add a test for this color so we can merge the fix and release?

File: ncscolor.js (hunk 1 above)
Comment (milahu) [reply ncscolor.js]: tested with my `compare-ncscolor-vs-w3color.js` and the result is the same : )

**Comment by milahu** [comment]
> add a test for this color

two seconds .. im working on a generalized test, to check for continuity / steadiness

**Comment by milahu** [comment]
[test-ncscolor.js](https://github.com/milahu/compare-ncscolor-vs-w3color/blob/master/test-ncscolor.js)

output with bug

```
$ node test-ncscolor.js
mismatch: degree 276 / 400 = ncs 0580-G76Y:
            last             -> curr (mismatch)
  w3color:  rgb(253,255, 36) -> rgb(255,255, 51)
  ncsColor: rgb(253,255, 36) -> rgb(357,357, 51)

unsteady: degree 275 -> 276 / 400 = ncs 0580-G75Y -> 0580-G76Y:
  last: rgb(253,255, 36). ncsColor.col.r = 253
  curr: rgb(357,357, 51). ncsColor.col.r = 357 = +41%
unsteady: degree 275 -> 276 / 400 = ncs 0580-G75Y -> 0580-G76Y:
  last: rgb(253,255, 36). ncsColor.col.g = 255
  curr: rgb(357,357, 51). ncsColor.col.g = 357 = +40%
unsteady: degree 275 -> 276 / 400 = ncs 0580-G75Y -> 0580-G76Y:
  last: rgb(253,255, 36). ncsColor.col.b = 36
  curr: rgb(357,357, 51). ncsColor.col.b = 51 = +6%

unsteady: degree 276 -> 277 / 400 = ncs 0580-G76Y -> 0580-G77Y:
  last: rgb(357,357, 51). ncsColor.col.r = 357
  curr: rgb(254,253, 36). ncsColor.col.r = 254 = -40%
unsteady: degree 276 -> 277 / 400 = ncs 0580-G76Y -> 0580-G77Y:
  last: rgb(357,357, 51). ncsColor.col.g = 357
  curr: rgb(254,253, 36). ncsColor.col.g = 253 = -41%
unsteady: degree 276 -> 277 / 400 = ncs 0580-G76Y -> 0580-G77Y:
  last: rgb(357,357, 51). ncsColor.col.b = 51
  curr: rgb(254,253, 36). ncsColor.col.b = 36 = -6%

errors found :(
```

the tolerances for rgb are around 6% -- 40% difference is obviously unsteady

but the w3color bug is not detected ... there the error is only 6%
but the differences jump like 0 - 0 - 1 - 0 - 0 - 6 - 0 - 0 - 1 - 0 ...
so that second-order unsteadiness can be detected

output after patch

```
$ node test-ncscolor.js 
mismatch: degree 276 / 400 = ncs 0580-G76Y:
            last             -> curr (mismatch)
  w3color:  rgb(253,255, 36) -> rgb(255,255, 51)
  ncsColor: rgb(253,255, 36) -> rgb(255,255, 36)

looking good :)
```

"looking good" cos mismatches are only warnings

the mismatch checking could be removed or deactivated

todo: test more nuances

**Comment by milahu** [comment]
> that second-order unsteadiness can be detected

```
$ node test-ncscolor.js 

unsteady diff:
  last: 275 -> 276 = 0580-G75Y -> 0580-G76Y: w3color.col.blue -= +6%
  curr: 276 -> 277 = 0580-G76Y -> 0580-G77Y: w3color.col.blue -= -6%' -> diff -12% (max 6%)
```

good news, looks like that was the only bug in w3color

**Comment by m90** [comment]
That's some serious color science you have going on there. Do you think it would make sense to port the test to this repo / this PR? I feel it'd make sense to run it by default as even if the output is an approximation, it should always be deterministic.

Let me know if you wanna do that or if I should merge as is. Thank you!

**Comment by milahu** [comment]
> I feel it'd make sense to run it by default

you mean as github action? sounds like overkill

i would just include the test script in this repo
for the next random stranger who wants to hack on this algorithm ...

**Comment by m90** [comment]
There is a very limited mocha test setup via `npm test` set up already so I guess that would be the easiest way to make it available. If you just want to add the script I can also take care of adapting it to work with the existing setup.

**Comment by milahu** [comment]
> I can also take care of adapting it to work with the existing setup.

cool : ) i hope the million lint errors are auto-fixable ... happy hacking

**Comment by m90** [comment]
Published as 1.0.1 - Thanks.
